Possibly undefined behavior in definition of Tag_val

The definition of Tag_val (little endian) is:

```
#define Tag_val(val) (((volatile unsigned char *) (val)) [-sizeof(value)])
```

I believe this code exhibits undefined behavior, specifically pointer overflow. Almost certainly this UB is harmless, but my company's internal compiler toolchain flags code using this macro when sanitizers are enabled.

My explanation for the UB is this:
1. sizeof(value) gives a size_t value, which is unsigned
2. negating the size_t is defined, but wraps (0xfffffffffffffff8)
3. adding this value to the pointer `val` overflows, although if the arithmetic wraps the resulting value is correct

I am not an expert here, but I believe this kind of pointer overflow is technically undefined behavior. Presumably no compiler exploits this for optimizations. I think this logic also relies on the pointer's storage and size_t having the same bit width.

It's not clear that this should be fixed in the OCaml source, since it's unlikely there is a real bug, and unlikely that compilers would exploit this UB in a way that breaks code. However, the false positive from sanitizers might be motivation enough. We can work around the UBsan issue by using `Tag_hd(Hd_val)` instead or by disabling sanitizers.

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

--- a/Changes
+++ b/Changes
@@ -522,6 +522,10 @@ Working version
   (Miod Vallat and Xavier Leroy, report by Jan Midtgaard, review by
    KC Sivaramakrishnan)
 
+- #12528, #12703: Avoid pointer arithmetic overflow in Tag_val macro
+  (very likely harmless, but can trigger alarms)
+  (Xavier Leroy, report by Sam Goldman, review by Guillaume Munch-Maccagnoni)
+
 - #12593: TSan should handle Effect.Unhandled correctly
   (Fabrice Buoro and Olivier Nicole, report by Jan Midtgaard and Miod Vallat,
    review by Gabriel Scherer)

--- a/runtime/caml/mlvalues.h
+++ b/runtime/caml/mlvalues.h
@@ -205,7 +205,7 @@ Caml_inline header_t Hd_val(value val)
 #define Tag_hp(hp) (((volatile unsigned char *) (hp)) [sizeof(value)-1])
                                                  /* Also an l-value. */
 #else
-#define Tag_val(val) (((volatile unsigned char *) (val)) [-sizeof(value)])
+#define Tag_val(val) (((volatile unsigned char *) (val)) [- (int)sizeof(value)])
                                                  /* Also an l-value. */
 #define Tag_hp(hp) (((volatile unsigned char *) (hp)) [0])
                                                  /* Also an l-value. */

